File: Практическая работа №3/src/main/java/io/github/ikbo0619/testing/team6/durak/Deck.java
package io.github.ikbo0619.testing.team6.durak;

import io.github.ikbo0619.testing.team6.durak.cardstatic.Static;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Stack;

public class Deck {
    private final Stack<Card> cards;

    public Deck() {
        cards = new Stack<>();

        ArrayList<Card> allCards = new ArrayList<>();
        for (String rank : Static.ranks)
            for (String suit : Static.suits) {
                allCards.add(new Card(rank, suit));
                Collections.shuffle(allCards);
            }

        for (Card card : allCards)
            cards.push(card);
    }

    public Card draw() {
        return isEmpty() ? null : cards.pop();
    }

    public boolean isEmpty() {
        return cards.empty();
    }

    public int size() {
        return cards.size();
    }

    public void reinsert(Card t) {
        cards.add(0, t);
    }

    @Override
    public String toString() {
        String ret = "[Bottom]\n";
        for (Card card : cards)
            ret += card + "\n";
        ret += "[Top]\n";
        return ret;
    }
}

File: Практическая работа №6/№3 Имеющаяся программная система/src/main/kotlin/Deck.kt
import java.util.*
import kotlin.concurrent.thread

class Deck {
    var cards = Stack<Card>()
        private set

    init {
        val allCards = ArrayList<Card>().apply {
            for (rank in Rank.values()) for (suit in Suits.values())
                add(Card(rank, suit))
            shuffle()
        }

        try { require(allCards.size == Rank.values().size * Suits.values().size) }
        catch (e: Exception) { println("Deck initialization failed!") }

        for (card in allCards)
            cards.push(card)

        thread(block = { while (true) {
            try { require(size() <= Rank.values().size * Suits.values().size) }
            catch (e: Exception) { println("Deck must not contain more cards than was initialized!") }
        }})
    }

    fun draw(): Card? {
        if (isEmpty())
            return null

        val sizePrev = size()
        val card = cards.pop()

        try { require(sizePrev > size()) }
        catch (e: Exception) { println("Drawing failed!") }

        return card
    }

    fun isEmpty() = cards.empty()

    fun size(): Int = cards.size

    fun reinsert(trump: Card) {
        val sizePrev = size()
        cards.add(0, trump)

        try { require(sizePrev < size()) }
        catch (e: Exception) { println("Reinsertion failed!") }
    }

    override fun toString() =
        buildString {
            append("[Bottom]\n")
            for (card in cards)
                append("$card\n")
            append("[Top]\n")
        }
}
